File: Lineal_structure/PY/Cola/Cola.py
from Nodo import Nodo
import os
import webbrowser

class Cola:
    def __init__(self):
        self.frente = None
        self.atras = None
        self.tamanio = 0

    def getTamanio(self):
        return self.tamanio

    def arribo(self, dato):
        nodo_nuevo = Nodo(dato)

        if self.frente is None:
            self.frente = nodo_nuevo
        else:
            self.atras.siguiente = nodo_nuevo

        self.atras = nodo_nuevo
        self.tamanio += 1

    def atencion(self):
        dato = self.frente.dato
        self.frente = self.frente.siguiente

        if self.frente is None:
            self.atras = None

        self.tamanio -= 1
        return dato

    def esVacio(self):
        return self.frente is None

    def en_frente(self):
        return self.frente.dato.__str__()

    def getTamanio(self):
        return self.tamanio

    def mover_al_final(self):
        dato = self.atencion()
        self.arribo(dato)
        return dato

    def listar(self):
        aux = self.frente

        if self.esVacio():
            return print("La cola esta vacia")

        self.temp = []

        while aux is not None:
            self.temp.append(aux.dato)
            aux = aux.siguiente

        print(self.temp)

    def graficarDot(self):
        temporal = self.frente
        cont = 0
        cadena = ""
        cadena += "digraph G { \n"
        cadena += "rankdir=RL \n"

        while temporal is not None:
            cadena += "Node" + str(cont) + "[label=\"" + temporal.dato+"\"];\n"

            if temporal is not self.frente:
                cadena += "Node" + str(cont) + " -> " + "Node" + str(cont - 1) + ";\n"

            temporal = temporal.siguiente
            cont += 1


        cadena += "}"
        #print(cadena)
        dot = "Lineal_structure\PY\Cola\Cola_Dot.txt"
        with open(dot, 'w') as grafo:
            grafo.write(cadena)
        result = "Lineal_structure\PY\Cola\Cola_Graph.png"
        os.system("dot -Tpng "+ dot +" -o "+ result)
        webbrowser.open(result)

File: Lineal_structure/Go/Cola/Cola.go
package main

import (
    "fmt"
    "os"
    "os/exec"
)

type Nodo struct {
    dato int
    siguiente *Nodo
}

type Cola struct {

    frente *Nodo
    atras *Nodo
    tamanio int
}

func (this *Cola) getTamanio() int {
    return this.tamanio
}

func (this *Cola) arribo(dato string) {

    nodo_nuevo := Nodo{dato, nil}

    if this.frente == nil {
        this.frente = &nodo_nuevo
    } else {
        this.atras.siguiente = &nodo_nuevo
	}

    this.atras = &nodo_nuevo
    this.tamanio += 1
}

func (this *Cola) atencion() string {

    dato := this.frente.dato
    this.frente = this.frente.siguiente

    if this.frente == nil {
    	 this.atras = nil
    }
    this.tamanio -= 1
    return dato
}

func (this *Cola) esVacio() bool {
    return this.frente == nil
}

func (this *Cola) en_frente() string {
    return this.frente.dato
}

func (this *Cola) getTamanio() int {
    return this.tamanio
}

func (this *Cola) mover_al_final() string {
    dato := this.atencion()
    this.arribo(dato)
    return dato
}

func (this *Cola) listar() {

    aux := this.frente

    if this.esVacio() {
        return fmt.Println("La cola esta vacia")
    }
    this.temp := []string{}
    for aux != nil {
        this.temp = append(this.temp, aux.dato)
        aux = aux.siguiente
    }
    fmt.Println(this.temp)
}



func (this *Cola) graficarDot() {

    temporal := this.frente
    cont := 0
    cadena := ""

    cadena += "digraph G { \n"
    cadena += "rankdir=RL \n"

    for temporal != nil {
        cadena += "Node" + strconv.Itoa(cont) + "[label=\"" + temporal.dato+"\"];\n"
        if temporal != this.frente {
            cadena += "Node" + strconv.Itoa(cont) + " -> " + "Node" + strconv.Itoa(cont - 1) + ";\n"
        }
        temporal = temporal.siguiente
        cont += 1
    }

    cadena += "}"
    //print(cadena)

    dot := "Lineal_structure\\Go\\Cola\\Cola_Dot.txt"
    with open(dot, 'w') as grafo:
        grafo.write(cadena)
    result := "Lineal_structure\\Go\\Cola\\Cola_Graph.png"
    os.system("dot -Tpng "+ dot +" -o "+ result)
    webbrowser.open(result)

}

func main() ()  {
    
}